tests/LeanMapperQuery/Entity.hasMany.phpt

```diff
@@ -0,0 +1,182 @@
+<?php
+
+use LeanMapper\Repository;
+use LeanMapperQuery\Entity;
+use LeanMapperQuery\Query;
+use Tester\Assert;
+
+require_once __DIR__ . '/../bootstrap.php';
+
+class BaseEntity extends Entity
+{
+	public function find($field, $query)
+	{
+		$entities = $this->queryProperty($field, $query);
+		return $this->entityFactory->createCollection($entities);
+	}
+}
+
+class BaseRepository extends Repository
+{
+	public function findAll()
+	{
+		return $this->createEntities($this->createFluent()->fetchAll());
+	}
+}
+
+class BookRepository extends BaseRepository
+{
+}
+
+/**
+ * @property int    $id
+ * @property string $name
+ */
+class Tag extends BaseEntity
+{
+}
+
+/**
+ * @property int         $id
+ * @property Tag[]       $tagsIn      m:hasMany
+ * @property-read Tag[]  $tagsUnion   m:hasMany(#union)
+ * @property string      $name
+ */
+class Book extends BaseEntity
+{
+}
+
+////////////////
+
+$sqls = [];
+$connection->onEvent[] = function ($event) use (&$sqls) {
+	$sqls[] = $event->sql;
+};
+$bookRepository = new BookRepository($connection, $mapper, $entityFactory);
+
+function extractTags(BookRepository $bookRepository, $tagProperty, Query $query)
+{
+	$result = [];
+
+	foreach ($bookRepository->findAll() as $book) {
+		$tags = $book->find($tagProperty, $query);
+
+		if (count($tags) <= 1) {
+			$tag = reset($tags);
+			$result[$book->id] = $tag ? $tag->name : NULL;
+
+		} else {
+			foreach ($tags as $tag) {
+				$result[$book->id][] = $tag->name;
+			}
+		}
+	}
+
+	return $result;
+}
+
+
+////////////////
+
+$query = getQuery()
+	->limit(1);
+
+$expected = [
+	1 => 'popular',
+	2 => NULL,
+	3 => 'ebook',
+	4 => 'popular',
+	5 => NULL,
+];
+
+$sqls = [];
+Assert::same($expected, extractTags($bookRepository, 'tagsIn', $query));
+Assert::same($expected, extractTags($bookRepository, 'tagsUnion', $query));
+Assert::same($sqls, [
+	'SELECT [book].* FROM [book]',
+	'SELECT * FROM (SELECT [book_tag].* FROM [book_tag] LEFT JOIN [tag] ON [book_tag].[tag_id] = [tag].[id] WHERE [book_tag].[book_id] = 1 LIMIT 1) UNION SELECT * FROM (SELECT [book_tag].* FROM [book_tag] LEFT JOIN [tag] ON [book_tag].[tag_id] = [tag].[id] WHERE [book_tag].[book_id] = 2 LIMIT 1) UNION SELECT * FROM (SELECT [book_tag].* FROM [book_tag] LEFT JOIN [tag] ON [book_tag].[tag_id] = [tag].[id] WHERE [book_tag].[book_id] = 3 LIMIT 1) UNION SELECT * FROM (SELECT [book_tag].* FROM [book_tag] LEFT JOIN [tag] ON [book_tag].[tag_id] = [tag].[id] WHERE [book_tag].[book_id] = 4 LIMIT 1) UNION SELECT * FROM (SELECT [book_tag].* FROM [book_tag] LEFT JOIN [tag] ON [book_tag].[tag_id] = [tag].[id] WHERE [book_tag].[book_id] = 5 LIMIT 1)',
```

Comment:
Že jsou to dotazy dlouhé, je politování hodné. Nevím, co s tím jiného. Teoreticky by se to dalo nějak přes `implode(' UNION ', ...)` zredukovat, ale asi to nestojí za to?

Reply:
Řekl bych, že to přehlednější je. Nevím, co s tím jiného. Nechme to tak.